src/covers/discogscoverprovider.cpp

```diff
@@ -0,0 +1,373 @@
+                                              int r_id) {
+  reply->deleteLater();
+
+  DiscogsCoverReleaseContext* r_ctx;
+  if (!requests_release_.contains(r_id)) {
+    return;
+  }
+  r_ctx = requests_release_.value(r_id);
+  if (r_ctx == nullptr) return;
+
+  DiscogsCoverSearchContext* s_ctx;
+  if (!requests_search_.contains(s_id)) {
+    EndSearch(r_ctx);
+    return;
+  }
+  s_ctx = requests_search_.value(s_id);
+  if (s_ctx == nullptr) return;
+
+  QString json_string = reply->readAll();
+  QByteArray json_bytes = json_string.toLocal8Bit();
+  auto json_doc = QJsonDocument::fromJson(json_bytes);
+  if (json_doc.isNull()) {
+    qLog(Error) << "Discogs: Failed to create JSON doc.";
+    EndSearch(s_ctx, r_ctx);
+    return;
+  }
+  if (!json_doc.isObject()) {
+    qLog(Error) << "Discogs: JSON is not an object.";
+    EndSearch(s_ctx, r_ctx);
+    return;
+  }
+
+  QJsonObject json_obj = json_doc.object();
+  if (json_obj.isEmpty()) {
+    qLog(Error) << "Discogs: JSON object is empty.";
+    EndSearch(s_ctx, r_ctx);
+    return;
+  }
+
+  QVariantMap reply_map = json_obj.toVariantMap();
+  if (!reply_map.contains("images")) {
+    EndSearch(s_ctx, r_ctx);
+    return;
+  }
+
+  QVariantList results = reply_map["images"].toList();
+
+  for (const QVariant& result : results) {
+    QVariantMap result_map = result.toMap();
+    CoverSearchResult cover_result;
+    cover_result.description = s_ctx->title;
+
+    if (result_map.contains("type")) {
+      QString type = result_map["type"].toString();
+      if (type != "primary") continue;
+    }
+    if (result_map.contains("resource_url")) {
+      cover_result.image_url = QUrl(result_map["resource_url"].toString());
+    }
+    if (cover_result.image_url.isEmpty()) continue;
+    s_ctx->results.append(cover_result);
+  }
+
+  EndSearch(s_ctx, r_ctx);
+}
+
+void DiscogsCoverProvider::SearchRequestError(QNetworkReply::NetworkError error,
+                                              QNetworkReply* reply, int s_id) {
+  DiscogsCoverSearchContext* s_ctx;
+  if (!requests_search_.contains(s_id)) {
+    return;
+  }
+  s_ctx = requests_search_.value(s_id);
+  if (s_ctx == nullptr) return;
+
+  EndSearch(s_ctx);
+}
+
+void DiscogsCoverProvider::ReleaseRequestError(
+    QNetworkReply::NetworkError error, QNetworkReply* reply, int s_id,
+    int r_id) {
+  DiscogsCoverReleaseContext* r_ctx;
+  if (!requests_release_.contains(r_id)) {
+    return;
+  }
+  r_ctx = requests_release_.value(r_id);
+  if (r_ctx == nullptr) return;
+
+  DiscogsCoverSearchContext* s_ctx;
+  if (!requests_search_.contains(s_id)) {
+    EndSearch(r_ctx);
+  }
+  s_ctx = requests_search_.value(s_id);
+  if (s_ctx == nullptr) return;
+
+  EndSearch(s_ctx, r_ctx);
+}
+
+void DiscogsCoverProvider::EndSearch(DiscogsCoverSearchContext* s_ctx,
+                                     DiscogsCoverReleaseContext* r_ctx) {
+  (void)requests_release_.remove(r_ctx->id);
```

Comment:
You don't need the `(void)`

Reply:
You probably just want to do: `delete requests_release_.take(r_ctx->id);` anyway

Reply:
Agree